### PR TITLE
fix: issue130 Return a 404 error while adding a pet to an unexisting owner 

### PR DESCRIPTION
**Pull request by earlspilner**
Improved handling of 404 errors for non-existing owners and pets in the `POST /owners/{ownerId}/pets` endpoint. Also switched to using records for error information to simplify and modernize the code.

This PR addresses https://github.com/spring-petclinic/spring-petclinic-rest/issues/130 and enhances both the error handling and the structure of the error responses.

**Changes:**

1. Fixed the endpoint to return a 404 Not Found status for non-existing owners or pets instead of a 400 Bad Request.
2. Refactored error handling by replacing the previous class-based error representation with Java record, which stands for a more concise and immutable way to handle error details.
3. Few test changes, because the old version was wrong

@arey, please review and merge this update.

### CODE REVIEW

**Comment by earlspilner** [comment]
Okay, finally this build was successful :)
I've made some changes in tests, because from my point there're mistakes

**Comment by arey** [comment]
Hi @earlspilner 
Thank your for your contribution
In the `openapi.yaml` file, mayve change the `Pet not found.` by `Owner not found`?
```
        404:
          description: Pet not found.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/RestError'
```

File: src/main/java/org/springframework/samples/petclinic/rest/advice/ExceptionControllerAdvice.java [hunk 1]
```diff
@@ -33,53 +34,76 @@
...
-            e1.printStackTrace();
+    /**
+     * Record for storing error information.
+     * <p>
+     * This record encapsulates the class name and message of the exception.
+     *
+     * @param className The name of the exception class
+     * @param exMessage The message of the exception
+     */
+    private record ErrorInfo(String className, String exMessage) {
```
Comment (arey) [inline src/main/java/org/springframework/samples/petclinic/rest/advice/ExceptionControllerAdvice.java:56]: praise: Thank you for using record

File: src/main/java/org/springframework/samples/petclinic/rest/advice/ExceptionControllerAdvice.java [hunk 2]
```diff
@@ -33,53 +34,76 @@
...
+    }
+
+    /**
+     * Handles {@link DataIntegrityViolationException} which typically indicates database constraint violations.
+     * This method returns a 404 Not Found status if an entity does not exist.
+     *
+     * @param ex The {@link DataIntegrityViolationException} to be handled
+     * @return A {@link ResponseEntity} containing the error information and a 404 Not Found status
+     */
+    @ExceptionHandler(DataIntegrityViolationException.class)
```
[the remark this answers is not in the document]
Comment (earlspilner) [reply src/main/java/org/springframework/samples/petclinic/rest/advice/ExceptionControllerAdvice.java:81]: answer: Yeah, I checked it out. The DataIntegrityViolationException pops up in all three profiles: jdbc, jpa, and springdatajpa. Everything’s running smooth.

**Comment by earlspilner** [comment]
> In the `openapi.yaml` file, mayve change the `Pet not found.` by `Owner not found`?

I hear you, but here’s the thing — sometimes the issue might not just be with the owner. It could be that neither the `ownerId` nor the `typeId` is found in the request. So if we just say "**Owner not found**," it might not cover all the bases. I’d suggest we go with something like "**Pet or Owner not found**" to make sure we’re addressing any possible issue that could come up.

**Comment by arey** [comment]
> "Pet or Owner not found"
I'm agree. I didn't think to the pet's type. Good catch.

**Comment by arey** [comment]
All it's ok. Thanks a lot. Feel free to contribute to any other petclinic project / issue.

**Pull request merged**